gp-templates/helper-functions.php

```diff
@@ -69,123 +69,150 @@ function gp_prepare_translation_textarea( $text ) {
+			return $translation;
+		}
+	} else {
+		// Build our glossary search.
+		$glossary_entries = $glossary->get_entries();
+		$cached_glossary = $glossary->id;
+		if ( empty( $glossary_entries ) ) {
+			$terms_search = false;
+			return $translation;
+		}
 
-	$glossary_entries_terms_array = array();
-	$glossary_term_reference      = array();
+		$glossary_entries_suffixes = gp_glossary_add_suffixes( $glossary_entries );
 
-	// Since glossary entries terms have been returned as an array of strings, we need to break it up and create back references.
-	foreach ( $glossary_entries_terms as $i => $terms ) {
-		// Split the terms in to an array and make sure they are all lower case for later use.
-		$term_list = explode( '|', $terms );
-		$term_list = array_map( 'strtolower', $term_list );
+		$glossary_entries_reference = array();
+		foreach ( $glossary_entries as $id => $value ) {
+			$term = strtolower( $value->term );
+			if ( ! isset( $glossary_entries_reference[ $term ] ) ) {
+				$glossary_entries_reference[ $term ] = $id;
+				continue;
+			}
 
-		foreach ( $term_list as $term ) {
-			// Add the term to the terms array.
-			$glossary_entries_terms_array[] = $term;
-			// Make a back reference to the term's glossary entry.
-			$glossary_term_reference[ $term ][] = $i;
+			if ( ! is_array( $glossary_entries_reference[ $term ] ) ) {
+				$glossary_entries_reference[ $term ] = array( $glossary_entries_reference[ $term ] );
+			}
+			$glossary_entries_reference[ $term ][] = $id;
 		}
-	}
 
-	// Sort glossary entries by word count, to search first for multiple word terms, words with hyphens or both.
-	usort(
-		$glossary_entries_terms_array,
-		function( $a, $b ) {
-			$a_words = preg_split( '/[\s-]/', $a );
-			$b_words = preg_split( '/[\s-]/', $b );
-			return count( $b_words ) <=> count( $a_words );
+		$terms_search = '\b(';
+		foreach ( $glossary_entries_suffixes as $term => $suffixes ) {
+			$terms_search .= preg_quote( $term, '/' );
+
+			if ( ! empty( $suffixes ) ) {
+				$terms_search .= '(?:' . implode( '|', $suffixes ) . ')?';
+			}
+
+			$terms_search .= '|';
+
+			$referenced_term = $term;
+			if ( ! isset( $glossary_entries_reference[ $referenced_term ] ) ) {
+				foreach ( $suffixes as $suffix ) {
+					if ( isset( $glossary_entries_reference[ $term . $suffix ] ) ) {
+						$referenced_term = $term . $suffix;
+					}
+				}
+				if ( ! isset( $glossary_entries_reference[ $referenced_term ] ) ) {
+					// This should not happen but we don't want to access a non existing item below.
+					continue;
+				}
+			}
+
+			$referenced_term = $glossary_entries_reference[ $referenced_term ];
+			// Add the suffixed terms to the lookup table.
+			foreach ( $suffixes as $suffix ) {
+				if ( isset( $glossary_entries_reference[ $term . $suffix ] ) ) {
+					$glossary_entries_reference[ $term . $suffix ] = array_values( array_unique( array_merge( (array) $referenced_term, (array) $glossary_entries_reference[ $term . $suffix ] ) ) );
```

Comment:
```suggestion
					$glossary_entries_reference[ $term . $suffix ] = array_values( array_unique( array_merge( (array) $glossary_entries_reference[ $term . $suffix ] ), (array) $referenced_term ) );
```


Switching the order here would address https://github.com/GlotPress/GlotPress/pull/1395#issuecomment-1090621335